src/rpc/rawtransaction.cpp

```diff
@@ -922,67 +924,77 @@ static RPCHelpMan testmempoolaccept()
-
-    UniValue result(UniValue::VARR);
-    UniValue result_0(UniValue::VOBJ);
-    result_0.pushKV("txid", tx->GetHash().GetHex());
-    result_0.pushKV("wtxid", tx->GetWitnessHash().GetHex());
+    std::vector<CTransactionRef> txns;
 
-    TxValidationState state;
-    bool test_accept_res;
-    CAmount fee{0};
-    {
-        LOCK(cs_main);
-        test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee);
+    for (unsigned int i = 0; i < raw_transactions.size(); ++i) {
+        CMutableTransaction mtx;
+        if (!DecodeHexTx(mtx, raw_transactions[i].get_str())) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+        }
+        txns.emplace_back(MakeTransactionRef(std::move(mtx)));
     }
 
-    // Check that fee does not exceed maximum fee
-    if (test_accept_res && max_raw_tx_fee && fee > max_raw_tx_fee) {
-        result_0.pushKV("allowed", false);
-        result_0.pushKV("reject-reason", "max-fee-exceeded");
-        result.push_back(std::move(result_0));
-        return result;
+    std::vector<MempoolAcceptResult> validation_results;
+    if (txns.size() == 1) {
+        validation_results.emplace_back(WITH_LOCK(cs_main,
+            return AcceptToMemoryPool(mempool, txns[0], false /* bypass_limits */, true /* test_accept */)));
+    } else {
+        validation_results = WITH_LOCK(cs_main, return ProcessNewPackage(mempool, txns, true));
     }
-    result_0.pushKV("allowed", test_accept_res);
 
-    // Only return the fee and vsize if the transaction would pass ATMP.
-    // These can be used to calculate the feerate.
-    if (test_accept_res) {
-        result_0.pushKV("vsize", virtual_size);
-        UniValue fees(UniValue::VOBJ);
-        fees.pushKV("base", ValueFromAmount(fee));
-        result_0.pushKV("fees", fees);
-    } else {
-        if (state.IsInvalid()) {
-            if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
-                result_0.pushKV("reject-reason", "missing-inputs");
+    auto tx_accepted = [](MempoolAcceptResult& res) { return res.m_accepted; };
+    const bool all_valid = std::all_of(validation_results.begin(), validation_results.end(), tx_accepted);
+    // ProcessNewPackage should return a MempoolAcceptResult per transaction
+    // or exactly 1 for the first error that occurs.
+    CHECK_NONFATAL(all_valid || validation_results.size() == 1);
+
+    // TODO: Report absurd fees for packages.
+
+    UniValue result(UniValue::VARR);
+    if (all_valid) {
+        for (unsigned int i = 0; i < validation_results.size(); ++i) {
+            UniValue result_inner(UniValue::VOBJ);
+            const CTransaction tx = *txns[i];
+            result_inner.pushKV("txid", tx.GetHash().GetHex());
+            result_inner.pushKV("wtxid", tx.GetWitnessHash().GetHex());
+
+            const CAmount fee = validation_results[i].m_base_fees.value();
```

Comment:
I don't think this local variable is needed. It's only used in one place below (in the if conditional). You could just use `validation_results[i].m_base_fees.value()` there.

Reply:
Also used in `fees.pushKV("base", ValueFromAmount(fee))` if that's better?

Reply:
Oops. You're right. Ignore this!